File: HashmapAndSet/destCity.js
const destCity = (paths) => {
  const hasOutgoing = new Set();
  for (let [origin, destination] of paths) {
    hasOutgoing.add(origin);
  }

  for (let [origin, destination] of paths) {
    if (!hasOutgoing.has(destination)) return destination;
  }

  return null; // In case there is no destination city
};

// Time Complexity: O(n) where n is the number of paths
// Space Complexity: O(n) for the set

File: HashmapAndSet/dest_city.py
from typing import List


class DestCity:
    def dest_city(self, paths: List[List[str]]) -> str:
        hasOutgoing = set()
        for origin, destination in paths:
            hasOutgoing.add(origin)
        for origin, destination in paths:
            if destination not in hasOutgoing:
                return destination
        return None

# Time complexity: O(n) where n is the number of paths.
# Space complexity: O(n) for the origins set.

paths = [["B", "C"], ["D", "B"], ["C", "A"]]
print(DestCity().dest_city(paths))
